mqt/qmap/subarchitectures.py

```diff
@@ -0,0 +1,275 @@
+"""
+Functionality for computing good subarchitectures for quantum circuit mapping.
+
+The function provided in this module... TODO
+"""
+from __future__ import annotations
+
+import os
+import pickle
+from collections import defaultdict
+from itertools import combinations
+from typing import Dict, List, NewType, Tuple, Union
+
+import networkx as nx
+import retworkx as rx
+
+Subarchitecture = NewType("Subarchitecture", Union[rx.PyGraph, List[Tuple[int, int]]])
+PartialOrder = NewType("PartialOrder", Dict[Tuple[int, int], Tuple[int, int]])
+
+package_directory = os.path.dirname(os.path.abspath(__file__))
+precomputed_backends = ["rigetti_16", "ibm_guadalupe_16", "sycamore_23"]
+
+
+class SubarchitectureOrder:
+    """
+    Class representing partial order of Subarchitectures.
+
+    Attributes
+    ----------
+    arch : Architecture
+        quantum computing architecture whose subarchitectures should be ordered
+    sgs: list[list[Architecture]]
+        subarchitectures of arch. sgs[i][j]: the j-th subarchitecture of size i
+    subarch_order : dict[tuple[int, int], list[tuple[int, int]]]
+        ordering of variables according to subarchitecture order.
+        tuples (n, i) are indices into self.sgs
+    desirable_subarchitectures: dict[tuple[int, int], list[tuple[int, int]]]
+        mapping of indices (n, i) to its desirable subarchitectures
+        tuples (n, i) are indices into self.sgs
+
+    Methods
+    -------
+    optimal_candidates(nqubits)
+        return optimal candidate for mapping quantum circuits of a given size
+    covering(nqubits, size)
+        return a covering for nqubit circuits limited by size
+    store_library(lib_name)
+        serialize this object to avoid recomputing the ordering in the future
+    """
+
+    def __init__(self, arch: Subarchitecture | str):
+        """
+        Initialize the partial order.
+
+        If an architecture is given, the order will be computed for this
+        specific architecture.
+        If a str is given instead, the ordering will be loaded from the
+        subarchitecture library of that name.
```

Comment:
Slightly confused on this one. Do you mean that the construction should happen in two different helper functions? I still can only define one constructor since Python doesn't support method overloading. Is there an idiom I am unaware of?

Reply:
Judging from https://stackoverflow.com/a/141777, `@classmethod`s could be a suitable option here.